### PR TITLE
Fix CA2022 warnings (Avoid inexact read with 'Stream.Read')

### PR DESCRIPTION
**Pull request by mpidash**
Follow-up from https://github.com/dotnet/roslyn-analyzers/pull/7208.

The following finding from the initial analyzer PR that flags an `UnmanagedMemoryStream ` was not fixed, as we excluded well known reliable stream types (`MemoryStream` and `UnmanagedMemoryStream`, see https://github.com/dotnet/roslyn-analyzers/issues/7268):

https://github.com/dotnet/runtime/blob/79dd9bae9bb881eb716b608577c4cedc6c9cba72/src/libraries/System.Private.CoreLib/src/System/IO/UnmanagedMemoryStreamWrapper.cs#L93

### CODE REVIEW

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.cs [hunk 1]
```diff
@@ -963,7 +963,7 @@ public string ReadExisting()
                 Buffer.BlockCopy(_inBuffer, _readPos, bytesReceived, 0, CachedBytesToRead);
             }
 
-            _internalSerialStream.Read(bytesReceived, CachedBytesToRead, bytesReceived.Length - (CachedBytesToRead));    // get everything
+            _internalSerialStream.ReadExactly(bytesReceived, CachedBytesToRead, bytesReceived.Length - (CachedBytesToRead));    // get everything
```
Comment (stephentoub) [inline src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:966]: This won't compile as written. This library builds for multiple target frameworks, some of which don't have ReadExactly.

File: src/libraries/System.ServiceModel.Syndication/src/System/ServiceModel/XmlBuffer.cs [hunk 2]
```diff
@@ -86,7 +86,7 @@ public void Close()
             _bufferState = BufferState.Reading;
             _buffer = new byte[_stream.Length];
             _stream.Position = 0;
-            _stream.Read(_buffer, 0, _buffer.Length);
+            _stream.ReadExactly(_buffer);
```
Comment (stephentoub) [inline src/libraries/System.ServiceModel.Syndication/src/System/ServiceModel/XmlBuffer.cs:89]: Same here

File: src/libraries/System.Speech/src/Internal/Synthesis/AudioBase.cs [hunk 3]
```diff
@@ -121,7 +121,7 @@ internal void PlayWaveFile(AudioData audio)
                     try
                     {
                         byte[] data = new byte[(int)audio._stream.Length];
-                        audio._stream.Read(data, 0, data.Length);
+                        audio._stream.ReadExactly(data);
```
Comment (stephentoub) [inline src/libraries/System.Speech/src/Internal/Synthesis/AudioBase.cs:124]: And here

File: src/libraries/System.Speech/src/Internal/Synthesis/EngineSite.cs [hunk 4]
```diff
@@ -174,7 +174,7 @@ public Stream LoadResource(Uri uri, string mediaType)
                     int cLen = (int)stream.Length;
                     MemoryStream memStream = new(cLen);
                     byte[] ab = new byte[cLen];
-                    stream.Read(ab, 0, ab.Length);
+                    stream.ReadExactly(ab);
```
Comment (stephentoub) [inline src/libraries/System.Speech/src/Internal/Synthesis/EngineSite.cs:177]: Basically all of them :)

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.cs (hunk 1 above)
Comment (mpidash) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:967]: Thanks, I've added a check for `NET7_0_OR_GREATER` before using `ReadExactly`.

Is there a build target to check locally if I've used a function that is not available in a target framework? I ran the tests locally with `.\build.cmd -subset clr+mono+libs+libs.tests -test -rc Release` but got no build error.

Comment (stephentoub) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:967]: Thanks. That will only end up building for the current platform. There used to be a command for building all flavors of all libraries, but I can't find it in the docs now. @ViktorHofer? In the meantime, you can cd into the src folder for the library and `dotnet build` will build all targets for it.

Comment (ViktorHofer) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:967]: `build.cmd libs -allconfigurations`

File: src/libraries/System.ServiceModel.Syndication/src/System/ServiceModel/XmlBuffer.cs [hunk 5]
```diff
@@ -86,7 +86,21 @@ public void Close()
...
             _stream.Position = 0;
-            _stream.Read(_buffer, 0, _buffer.Length);
+
+#if NET7_0_OR_GREATER
+            _stream.ReadExactly(_buffer);
+#else
+            int totalRead = 0;
+            while (totalRead < _stream.Length)
+            {
+                int bytesRead = _stream.Read(_buffer, totalRead, _stream.Length - totalRead);
```
Comment (buyaa-n) [inline src/libraries/System.ServiceModel.Syndication/src/System/ServiceModel/XmlBuffer.cs:96]: FYI build failure: `error CS1503: (NETCORE_ENGINEERING_TELEMETRY=Build) Argument 3: cannot convert from 'long' to 'int'`

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.cs [hunk 6]
```diff
@@ -963,7 +963,21 @@ public string ReadExisting()
                 Buffer.BlockCopy(_inBuffer, _readPos, bytesReceived, 0, CachedBytesToRead);
             }
 
-            _internalSerialStream.Read(bytesReceived, CachedBytesToRead, bytesReceived.Length - (CachedBytesToRead));    // get everything
+#if NET7_0_OR_GREATER
+            _internalSerialStream.ReadExactly(bytesReceived, CachedBytesToRead, bytesReceived.Length - (CachedBytesToRead));    // get everything
```
Comment (buyaa-n) [inline src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.cs:967]: NIT:
```suggestion
            _internalSerialStream.ReadExactly(bytesReceived, CachedBytesToRead, bytesReceived.Length - CachedBytesToRead);    // get everything
```

File: src/libraries/System.Speech/src/Internal/Synthesis/AudioBase.cs [hunk 7]
```diff
@@ -121,7 +121,22 @@ internal void PlayWaveFile(AudioData audio)
...
                         byte[] data = new byte[(int)audio._stream.Length];
-                        audio._stream.Read(data, 0, data.Length);
+
+#if NET7_0_OR_GREATER
+                        audio._stream.ReadExactly(data);
+#else
+                        int totalRead = 0;
+                        while (totalRead < audio._stream.Length)
+                        {
+                            int bytesRead = audio._stream.Read(data, totalRead, audio._stream.Length - totalRead);
```
Comment (buyaa-n) [inline src/libraries/System.Speech/src/Internal/Synthesis/AudioBase.cs:131]: NIT:
```suggestion
                        while (totalRead < data.Length)
                        {
                            int bytesRead = audio._stream.Read(data, totalRead, data.Length - totalRead);
```

Comment (mpidash) [reply src/libraries/System.Speech/src/Internal/Synthesis/AudioBase.cs:131]: Thanks, I will use `data.Length` here, but I am not sure why this does not trigger CS1503 as well.

**Review by buyaa-n — APPROVED** [review]
LGTM, thanks!

**Pull request merged**